src/main/java/com/commercetools/sync/producttypes/helpers/AttributeDefinitionReferenceResolver.java

```diff
@@ -0,0 +1,123 @@
+import java.util.Optional;
+import java.util.concurrent.CompletionStage;
+
+import static io.sphere.sdk.utils.CompletableFutureUtils.exceptionallyCompletedFuture;
+import static java.lang.String.format;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+public class AttributeDefinitionReferenceResolver
+    extends BaseReferenceResolver<AttributeDefinitionDraft, ProductTypeSyncOptions> {
+    private ProductTypeService productTypeService;
+
+
+    AttributeDefinitionReferenceResolver(@Nonnull final ProductTypeSyncOptions options,
+                                         @Nonnull final ProductTypeService productTypeService) {
+        super(options);
+        this.productTypeService = productTypeService;
+    }
+
+    /**
+     * Given an {@link AttributeDefinitionDraft} this method attempts to resolve the ProductType references, which can
+     * be exist on attributeDefinition with an AttributeType: NestedType or SetType of NestedType, to return a
+     * {@link CompletionStage} which contains a new instance of the draft with the resolved references.
+     * The keys of the references are taken from the id field of the references.
+     *
+     * @param attributeDefinitionDraft the attributeDefinitionDraft to resolve its references.
+     * @return a {@link CompletionStage} that contains as a result a new attributeDefinitionDraft instance with resolved
+     *         references or if there is no productType existing with the given key the draft will be returned as is
+     *         without the reference resolved. In case an error occurs during reference resolution, a
+     *         {@link ReferenceResolutionException}.
+     */
+    @Nonnull
+    public CompletionStage<AttributeDefinitionDraft> resolveReferences(
+        @Nonnull final AttributeDefinitionDraft attributeDefinitionDraft) {
+
+        final AttributeDefinitionDraftBuilder draftBuilder =
+            AttributeDefinitionDraftBuilder.of(attributeDefinitionDraft);
+
+        return resolveNestedTypeReferenceOrSetOfNested(draftBuilder)
+            .handle(ImmutablePair::new)
+            .thenCompose(result -> {
+                final Throwable exception = result.getValue();
+                final AttributeDefinitionDraftBuilder resolvedBuilder = result.getKey();
+                if (exception == null) {
+                    return completedFuture(resolvedBuilder.build());
+                } else {
+                    final String errorMessage =
+                        format("Failed to resolve references on attribute definition with name '%s'.",
+                            attributeDefinitionDraft.getName());
+                    return exceptionallyCompletedFuture(
+                        new ReferenceResolutionException(errorMessage, exception.getCause()));
+                }
+            });
+
+    }
+
+    @Nonnull
+    private CompletionStage<AttributeDefinitionDraftBuilder> resolveNestedTypeReferenceOrSetOfNested(
```

Comment:
it looks like it's the main logic of the resolveReferences, maybe simply name it resolveReferences ?

Reply:
649694b